tests/cppunit/status_test.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <status.h>
+
+TEST(StatusOr, Scalar) {
+  auto f = [](int x) -> StatusOr<int> {
+    if (x > 10) {
+        return {Status::NotOK, "x large than 10"};
+    }
+
+    return 2 * x + 5;
+  };
+
+  ASSERT_EQ(*f(1), 7);
+  ASSERT_EQ(*f(5), 15);
+  ASSERT_EQ(f(7).GetValue(), 19);
+  ASSERT_EQ(f(7).GetCode(), Status::cOK);
+  ASSERT_EQ(f(7).Msg(), "ok");
+  ASSERT_TRUE(f(6));
+  ASSERT_EQ(f(11).GetCode(), Status::NotOK);
+  ASSERT_EQ(f(11).Msg(), "x large than 10");
```

Comment:
what would happen if calling `f(11).GetValue()`?

Reply:
If you look at the definition of `StatusOr<T>::GetValue`:

```c++
  value_type& GetValue() & {
    CHECK(*this);
    return getValue();
  }

  value_type&& GetValue() && {
    CHECK(*this);
    return std::move(getValue());
  }

  const value_type& GetValue() const& {
    CHECK(*this);
    return getValue();
  }
```

You will find there is three different overload of this method via changing the qualifier of `*this`.

Hence, in `f(x).GetValue()`, obviously `f(x)` is a [prvalue](https://en.cppreference.com/w/cpp/language/value_category#prvalue) since a temporary object is returned, and then the signature `value_type&& GetValue() &&` is decided to be called after the [overload resolution](https://en.cppreference.com/w/cpp/language/overload_resolution), and thus you get an [xvalue](https://en.cppreference.com/w/cpp/language/value_category#xvalue) (which is a special kind of rvalue) referenced to the underlying value of the returned `StatusOr`.

Reply:
BTW, The classic usage of `StatusOr<T>` we expected is:
```
// 1
StatusOr<T> someProcess(...); // as a return type

// 2
T val;
...
return val; // return val typed T to StatusOr<T>

// 3
return {Status::someErrCode, "msg..."} // return an error status

// 4
auto res = someProcess(...); // get result typed StatusOr<T>
if(!res) return res; // forward it if it is an error status

doSomething(*res); // process the value inside the StatusOr
```

Reply:
i want to describe `11` is greater than `10`, `GetValue()` will return error?

Reply:
`GetValue()` will trigger a fatal error.
It is just like you call `f` as `Status f(int *output)`, but do not check the returned `Status` and directly dereference the `output`.

Reply:
I think that guarded by `CHECK(*this)`, it can cause a fatal error.

I wonder if we can generate a compile time error instead of runtime fatal in this case.

Reply:
Confirm its failure as:

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
F20220812 15:30:19.518088 1124890 status.h:202] Check failed: *this 
*** Check failure stack trace: ***
```

Reply:
Hi @tisonkun, thanks for your review.

> I wonder if we can generate a compile time error instead of runtime fatal in this case.

Good idea. We consider an example for this:
```c++
auto res = someProcess(...); // get result typed StatusOr<T>
// we cannot deref `res` here

if(!res) { 
  processError(res.GetCode(), res.Msg()); /* or */ return res; // we cannot deref `res` here
}

doSomething(*res); // we can deref it
```

In this example, only the `doSomething` part can actually dereference the `res`. 
We cannot dereference it both before `if(!res)` and inside `if(!res)`.
So only in the execution path with the condition `res.IsOk() == true` being satisfied, we can dereference `res`,

It is hardly to express in compile time, since it is related to the control flow.
Like a raw/smart pointer, we should check it does point to a value (not null) before dereference it.

Reply:
> checks whether a type like StatusOr is dereferenced after checking IsOk() using [symbolic execution](https://en.wikipedia.org/wiki/Symbolic_execution)

Sounds good tome.